### ISSUE TITLE
Compiler never terminates for simple GPU compilation

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#26029]
A user is trying to compile the following code:

```chapel
var x : complex(64);
x = 1.0 + 1.0i;
writeln(x, " ", abs(x));
```

and finds that the compilation seems to spin forever with Chapel 2.2

(Potentially) Salient details:

* This is a GPU compilation:
```
CHPL_LOCALE_MODEL=gpu
CHPL_GPU=nvidia
CHPL_GPU_ARCH=sm_70
```

* This is using the bundled version of LLVM:
```
CHPL_LLVM=bundled

$ chpl --version
warning: The prototype GPU support implies --no-checks. This may impact debuggability. To suppress this warning, compile with
--no-checks explicitly
chpl version 2.2.0
  built with LLVM version 18.1.6
  available LLVM targets: x86-64, x86, nvptx64, nvptx
Copyright 2020-2024 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```

The hang _seems_ to occur in the compiler's invocation of `fatbinary`:

```
$ chpl --print-commands testit.chpl
...
# object file to fatbinary
fatbinary -64 --create ...
[Ctrl-C] hit here
```

(though it could be that this step completed successfully and that the hang occurred within the compiler after this step, but before we'd printed something else).

* The CUDA version is 12.5 as set by CHPL_CUDA_PATH.
* The system version of CUDA is 12.6, which is in the path by default.

This suggests that maybe the compiler is using the system versions of `ptxas` and `fatbinary`, and that doing so could cause an incompatibility with other aspects of 12.5 that we use?

**Comment by e-kayrakli** [comment]
> The CUDA version is 12.5 as set by CHPL_CUDA_PATH.
> The system version of CUDA is 12.6, which is in the path by default.

This is certainly not great. Because:

- How can `CHPL_CUDA_PATH` get set wrong? It just asks nvcc about the version and path info?
- Mismatches like that created weird issues for us in the past. So, definitely a suspect.

Can `CHPL_CUDA_PATH` be set the 12.6's path and Chapel be rebuilt with that?

**Comment by bradcray** [comment]
> This suggests that maybe the compiler is using the system versions of ptxas and fatbinary, and that doing so could cause an incompatibility with other aspects of 12.5 that we use?

I asked the user to put the CUDA 12.5 path first in their path to check this theory, but the result was the same:

```
$ PATH=/usr/local/cuda-12.5//bin:$PATH
$ echo $PATH
/usr/local/cuda-12.5//bin:/usr/public/opt/CHAPEL/chapel-2.2.0_ugpu1/bin/linux64-x86_64:…etc.
$ chpl --print-commands testit.chpl
...
# object file to fatbinary
fatbinary -64 --create ...
[Ctrl-C] hit here
```

Also interesting:  Before hitting Ctrl-C, they checked processes running under their uid and did not find any mention of `fatbinary`.  So perhaps this was accurate:

> (though it could be that this step completed successfully and that the hang occurred within the compiler after this step, but before we'd printed something else).

Considering having them do a debug build of the compiler to get more information about where we are.

**Comment by bradcray** [comment]
> Can `CHPL_CUDA_PATH` be set the 12.6's path and Chapel be rebuilt with that?

That's on the list of things to try as well.

**Comment by e-kayrakli** [comment]
> (though it could be that this step completed successfully and that the hang occurred within the compiler after this step, but before we'd printed something else).

If this is the case, which is likely, they'd need a runtime rebuild. Our runtime links with CUDA libraries.

**Comment by bradcray** [comment]
User built a debug version of the compiler and tried to gdb it using `set follow-fork-mode child` and when it got to the hang, they hit ctrl-C but didn't get much useful information:

```
(gdb) info threads
No threads.
(gdb) where
No stack.
```

They then checked to see what processes were running and found the following two:

* `chpl --print-passes --print-commands bug_cmplx.chpl`
* `chpl --driver-compilation-phase --driver-tmp-dir /tmp/chpl-cfreese.deleteme-4Pc0BM --print-passes --print-commands bug_cmplx.chpl`

So they attached to the other process and got the following stack trace:

```
#0  0x0000000000571eb9 in isModuleSymbol (a=0x7fffeea46d80)
    at /path/to/compiler/include/baseAST.h:349
#1  0x00000000005720c4 in toModuleSymbol (a=0x7fffeea46d80)
    at /path/to/compiler/include/baseAST.h:406
#2  0x000000000057a728 in BaseAST::getModule (this=0x7fffeea46d80)
    at /path/to/compiler/AST/baseAST.cpp:392
#3  0x000000000057a7f9 in BaseAST::getModule (this=0x7ffff10c1c80)
    at /path/to/compiler/AST/baseAST.cpp:405
#4  0x0000000000a2a049 in findLocationIgnoringInternalInlining (cur=0x7ffff0588020)
    at /path/to/compiler/util/astlocs.cpp:196
#5  0x0000000000a3398f in printErrorHeader (ast=0x7ffff14a3ee0, astloc=...)
    at /path/to/compiler/util/misc.cpp:630
#6  0x0000000000a34794 in vhandleError(const BaseAST *, astlocT, const char *, typedef __va_list_tag __va_list_tag *) (ast=0x7ffff14a3ee0, astloc=...,
    fmt=0x54d9200 "Could not find C function for %s;  perhaps it is missing or is a macro?",
    args=0x7fffffff8e88)
    at /path/to/compiler/util/misc.cpp:960
#7  0x0000000000a344ea in handleError (ast=0x7ffff14a3ee0,
    fmt=0x54d9200 "Could not find C function for %s;  perhaps it is missing or is a macro?")
    at /path/to/compiler/util/misc.cpp:898
[and then a bunch of codegen/codegenDef calls that I'm leaving out
```

It's not obvious to me offhand why that would either hang or be involved in an infinite loop that didn't print out the error message.

I also want to do a quick check on something from those in the know (@e-kayrakli / @jabraham17 ):  Is it surprising that we'd have a process in this codegen stage after (or during) the invocation of the `fatbinary` step?  I would've thought that we'd be done with codegen before the call lto `fatbinary` (but that's just based on a gut reaction, no deep knowledge).

**Comment by bradcray** [comment]
Oh yeah, user is also willing to do an interactive session with someone if anyone is available.  I don't know that I have the LLVM+GPU+compiler driver chops to be that useful myself.

**Comment by bradcray** [comment]
Asking them to ctrl-C a few more times and print some more stack traces suggests that something may be spinning in the `findLocationIgnoringInternalInlining()` downwards part of the stack trace.  It also sounds like this hang may be specific to programs failing due to attempts to use complex on GPUs similar to #26019 (?).

**Comment by jabraham17** [comment]
I can reproduce this on a testing system with CUDA 12.4 and Chapel main. This is highly related to https://github.com/chapel-lang/chapel/issues/26019. Basically, the compiler is trying to report a nice error for the fact that `cabs` is missing with GPU compilation, and is stuck in an infinite loop in `findLocationIgnoringInternalInlining` (there is a literal `while (true)` :( ).

This seems to be caused by calling a missing C function from a standard module and can be replicated without GPUs by adding the following to a standard module and then trying to use it

```chapel
// in standard/Math.chpl
]
  proc foobar(x: int) {
    extern proc call_foobar(x: int): int;
    return call_foobar(x);
  }
```

`call_foobar` does not exist anywhere, and so the following code triggers the infinite loop

```chapel
use Math;
foobar(10);
```


But, if `foobar` is defined in a user module (not bundled/standard module), then there is no issue and the compiler correctly reports the error that `call_foobar` is missing.


So resolving https://github.com/chapel-lang/chapel/issues/26019 will resolve this case in particular, but it will not address the root cause.

**Comment by jabraham17** [comment]
I believe https://github.com/chapel-lang/chapel/pull/26037 will fix the root issue here. The code from the OP will still not work (because of https://github.com/chapel-lang/chapel/issues/26019), but it should now give the proper error at the proper line number, without hanging.

**Comment by e-kayrakli** [comment]
Thanks for the diagnosis and the patch, Jade!

> OP will still not work (because of https://github.com/chapel-lang/chapel/issues/26019)

Hopefully, #26019 will be fixed for 2.3 and that the OP will work in the next release :)
